File: terrarun/workspace.py
# Copyright (C) Matt Comben - All Rights Reserved
# Unauthorized copying of this file, via any medium is strictly prohibited
# Proprietary and confidential

import re
import sqlalchemy
import sqlalchemy.orm

from terrarun.base_object import BaseObject
from terrarun.config import Config
import terrarun.organisation
from terrarun.permissions.workspace import WorkspacePermissions
import terrarun.run
import terrarun.configuration
from terrarun.database import Base, Database


class Workspace(Base, BaseObject):

    ID_PREFIX = 'ws'
    RESERVED_NAMES = ['setttings']
    MINIMUM_NAME_LENGTH = 3

    __tablename__ = 'workspace'
    id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
    name = sqlalchemy.Column(sqlalchemy.String)
    organisation_id = sqlalchemy.Column(sqlalchemy.ForeignKey("organisation.id"), nullable=False)
    organisation = sqlalchemy.orm.relationship("Organisation", back_populates="workspaces")

    state_versions = sqlalchemy.orm.relation("StateVersion", back_populates="workspace")
    configuration_versions = sqlalchemy.orm.relation("ConfigurationVersion", back_populates="workspace")

    auto_apply = sqlalchemy.Column(sqlalchemy.Boolean, default=False)

    team_accesses = sqlalchemy.orm.relationship("TeamWorkspaceAccess", back_populates="workspace")

    _latest_state = None
    _latest_configuration_version = None
    _latest_run = None

    @classmethod
    def get_by_organisation_and_name(cls, organisation, workspace_name):
        """Return organisation object, if it exists within an organisation, by name."""
        session = Database.get_session()
        ws = session.query(Workspace).filter(
            Workspace.name==workspace_name,
            Workspace.organisation==organisation
        ).first()

        if not ws and Config().AUTO_CREATE_WORKSPACES:
            ws = Workspace.create(organisation=organisation, name=workspace_name)
        return ws

    @classmethod
    def validate_new_name(cls, organisation, name):
        """Ensure organisation does not already exist and name isn't reserved"""
        session = Database.get_session()
        existing_workspace = session.query(cls).filter(cls.name==name, cls.organisation==organisation).first()
        if existing_workspace:
            return False
        if name in cls.RESERVED_NAMES:
            return False
        if len(name) < cls.MINIMUM_NAME_LENGTH:
            return False
        if not re.match(r'^[a-zA-Z0-9-_]+$', name):
            return False
        return True

    @classmethod
    def create(cls, organisation, name):
        ws = cls(organisation=organisation, name=name)
        session = Database.get_session()
        session.add(ws)
        session.commit()
        return ws

    @property
    def runs(self):
        """Return runs for workspace"""
        runs = []
        for cv in self.configuration_versions:
            runs += cv.runs
        return runs

    @property
    def latest_configuration_version(self):
        """Return latest configuration version."""
        if self.configuration_versions:
            return self.configuration_versions[-1]
        return None

    @property
    def latest_state(self):
        """Return latest state version"""
        if self.state_versions:
            return self.state_versions[-1]
        return None

    @property
    def latest_run(self):
        """Return latest state version"""
        session = Database.get_session()
        run = session.query(
            terrarun.run.Run
        ).join(
            terrarun.configuration.ConfigurationVersion
        ).filter(
            terrarun.configuration.ConfigurationVersion.workspace == self
        ).order_by(
            terrarun.run.Run.created_at.desc()
        ).first()
        return run

    def get_api_details(self, effective_user):
        """Return details for workspace."""
        workspace_permissions = WorkspacePermissions(current_user=effective_user, workspace=self)
        return {
            "attributes": {
                "actions": {
                    "is-destroyable": True
                },
                "allow-destroy-plan": True,
                "apply-duration-average": 158000,
                "auto-apply": False,
                "auto-destroy-at": None,
                "created-at": "2021-06-03T17:50:20.307Z",
                "description": "An example workspace for documentation.",
                "environment": "default",
                "execution-mode": "remote",
                "file-triggers-enabled": True,
                "global-remote-state": False,
                "latest-change-at": "2021-06-23T17:50:48.815Z",
                "locked": False,
                "name": self.name,
                "operations": True,
                "permissions": workspace_permissions.get_api_permissions(),
                "plan-duration-average": 20000,
                "policy-check-failures": None,
                "queue-all-runs": False,
                "resource-count": 0,
                "run-failures": 6,
                "source": "terraform",
                "source-name": None,
                "source-url": None,
                "speculative-enabled": True,
                "structured-run-output-enabled": False,
                "terraform-version": "0.15.3",
                "trigger-prefixes": [],
                "updated-at": "2021-08-16T18:54:06.874Z",
                "vcs-repo": None,
                "vcs-repo-identifier": None,
                "working-directory": None,
                "workspace-kpis-runs-count": 7
            },
            "id": self.api_id,
            "links": {
                "self": f"/api/v2/organizations/{self.organisation.name}/workspaces/{self.name}"
            },
            "relationships": {
                "agent-pool": {
                    "data": {
                        "id": "apool-QxGd2tRjympfMvQc",
                        "type": "agent-pools"
                    }
                },
                "current-configuration-version": {
                    "data": {
                        "id": self.latest_configuration_version.api_id,
                        "type": "configuration-versions"
                    },
                    "links": {
                        "related": f"/api/v2/configuration-versions/{self.latest_configuration_version.api_id}"
                    }
                } if self.latest_configuration_version else {},
                "current-run": {
                    "data": {
                        "id": "run-UyCw2TDCmxtfdjmy",
                        "type": "runs"
                    },
                    "links": {
                        "related": "/api/v2/runs/run-UyCw2TDCmxtfdjmy"
                    }
                },
                "current-state-version": {
                    "data": {
                        "id": self.latest_state.api_id,
                        "type": "state-versions"
                    },
                    "links": {
                        "related": f"/api/v2/workspaces/{self.api_id}/current-state-version"
                    }
                } if self.latest_state else {},
                "latest-run": {
                    "data": {
                        "id": self.latest_run.api_id,
                        "type": "runs"
                    },
                    "links": {
                        "related": f"/api/v2/runs/{self.latest_run.api_id}"
                    }
                } if self._latest_run else {},
                "organization": {
                    "data": {
                        "id": self.organisation.name,
                        "type": "organizations"
                    }
                },
                "outputs": {
                    "data": []
                },
                "readme": {
                    "data": {
                        "id": "227247",
                        "type": "workspace-readme"
                    }
                },
                "remote-state-consumers": {
                    "links": {
                        "related": "/api/v2/workspaces/ws-qPhan8kDLymzv2uS/relationships/remote-state-consumers"
                    }
                }
            },
            "type": "workspaces"
        }

File: ui/src/app/interfaces/workspace.ts
export interface WorkspaceAttributes {
    "actions": {
        "is-destroyable": boolean;
    };
    "allow-destroy-plan": boolean;
    "apply-duration-average": number;
    "auto-apply": boolean;
    "auto-destroy-at": string | null;
    "created-at": string;
    "description": string | null;
    "environment": string | null;
    "execution-mode": string;
    "file-triggers-enabled": boolean;
    "global-remote-state": boolean
    "latest-change-at": string;
    "locked": boolean;
    "name": string;
    "operations": boolean;
    "permissions": {
        "can-create-state-versions": boolean;
        "can-destroy": boolean;
        "can-force-unlock": boolean;
        "can-lock": boolean;
        "can-manage-run-tasks": boolean;
        "can-manage-tags": boolean;
        "can-queue-apply": boolean;
        "can-queue-destroy": boolean;
        "can-queue-run": boolean;
        "can-read-assessment-result": boolean;
        "can-read-settings": boolean;
        "can-read-state-versions": boolean;
        "can-read-variable": boolean;
        "can-unlock": boolean;
        "can-update": boolean;
        "can-update-variable": boolean;
    };
    "plan-duration-average": number | null;
    "policy-check-failures": number | null;
    "queue-all-runs": boolean;
    "resource-count": number;
    "run-failures": number;
    "source": string;
    "source-name": string | null;
    "source-url": string | null;
    "speculative-enabled": boolean;
    "structured-run-output-enabled": boolean;
    "terraform-version": string | null;
    "trigger-prefixes": string[];
    "trigger-patterns": string[];
    "updated-at": string;
    "vcs-repo": {
        "branch": string | null;
        "ingress-submodules": string | null;
        "tags-regex": string | null;
        "identifier": string | null;
        "display-identifier": string | null;
        "oauth-token-id": string | null;
        "webhook-url": string | null;
        "repository-http-url": string | null;
        "service-provider": string | null;
    } | null;
    "vcs-repo-identifier": string | null;
    "working-directory": string | null;
    "workspace-kpis-runs-count": number;

    "overrides": {
        "allow-destroy-plan": boolean | null;
        "auto-apply": boolean | null;
        "execution-mode": string | null;
        "global-remote-state": boolean | null;
        "operations": string | null;
        "queue-all-runs": boolean | null;
        "speculative-enabled": boolean | null;
        "terraform-version": string | null;
    };
}

// Interface for workspace attributes when updating
// a workspace, allowing attributes to be nulled to revert
// them back to the project configuration
export interface WorkspaceUpdateAttributes extends Omit<WorkspaceAttributes, "queue-all-runs"> {
    "queue-all-runs": boolean | null;
}

export interface WorkspaceRelationships {
    "agent-pool": {
        "data": {
            "id": string;
            "type": string;
        }
    } | {};
    "current-configuration-version": {
        "data": {
            "id": string;
            "type": string;
        },
        "links": {
            "related": string;
        }
    } | {};
    "current-run": {
        "data": {
            "id": string;
            "type": string;
        },
        "links": {
            "related": string;
        }
    } | {};
    "project": {
        "data": {
            "id": string;
            "type": string;
        },
        "links": {
            "related": string;
        }
    },
    "current-state-version": {
        "data": {
            "id": string;
            "type": string;
        },
        "links": {
            "related": string;
        }
    } | {};
    "latest-run": {
        "data": {
            "id": string;
            "type": string;
        },
        "links": {
            "related": string;
        }
    } | {};
    "organization": {
        "data": {
            "id": string;
            "type": string;
        }
    };
    "outputs": {
        "data": {
            "id": string;
            "type": string;
        }[]
    };
    "tags": {
        "data": {
            "id": string;
            "type": string;
        }[];
    };
    "readme": {
        "data": {
            "id": string;
            "type": string;
        }
    };
    "remote-state-consumers": {
    };
}
